providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+    EXPORT_CB_ARG sub_arg;
+    int selection;
+    size_t publen = key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes;
+    size_t prvlen = key->minfo->prvkey_bytes + key->xinfo->prvkey_bytes;
+
+    /* The reported "bit" count is those of the ML-KEM key */
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, key->minfo->bits))
+            return 0;
+
+    /* The reported security bits are those of the ML-KEM key */
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, key->minfo->secbits))
+            return 0;
+
+    /* The ciphertext sizes are additive */
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, key->minfo->ctext_bytes + key->xinfo->pubkey_bytes))
+            return 0;
+
+    if (!mlx_kem_have_pubkey(key))
+        return 1;
+    pub = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (pub != NULL) {
+        if (pub->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        pub->return_size = publen;
+        if (pub->data == NULL)
+            pub = NULL;
+        else if (pub->data_size < publen)
+            return 0;
+    }
+    if (mlx_kem_have_prvkey(key)) {
+        prv = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+        if (prv) {
+            if (prv->data_type != OSSL_PARAM_OCTET_STRING)
+                return 0;
+            prv->return_size = prvlen;
+            if (prv->data == NULL)
+                prv = NULL;
+            else if (prv->data_size < prvlen)
+                return 0;
+        }
+    }
+    if (pub == NULL && prv == NULL)
+        return 1;
+
+    sub_arg.pubenc = pub ? pub->data : NULL;
+    sub_arg.prvenc = prv ? prv->data : NULL;
+    sub_arg.pubcount = 0;
+    sub_arg.prvcount = 0;
+
+    selection = prv == NULL ? 0 : OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    selection |= pub == NULL ? 0 : OSSL_KEYMGMT_SELECT_PUBLIC_KEY;
+    if (key->xinfo->group_name)
+        selection |= OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS;
+
+    /* Extract sub-component key material */
+    if (!export_sub(&sub_arg, selection, key))
+        return 0;
+
+    if ((pub != NULL && sub_arg.pubcount != 2)
+        || (prv != NULL && sub_arg.prvcount != 2))
+        return 0;
+
+    return 1;
+}
+
+static const OSSL_PARAM *mlx_kem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int mlx_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    MLX_KEY *key = vkey;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
```

Comment:
Why? Not following why no encoded-public-key is an immediate success return ...

Reply:
Nothing to do if no public key.